### PR TITLE
Mattermost.js refactor

### PR DESCRIPTION
**Pull request by miguelespinoza**
#### Summary
First changes for reducing cold startup time.
The goal is to create a lean entry point for JS Code for faster JS compilation.

This required breaking mattermost.js into:
* mattermost.android.js: initliaze listeners and export variables that can be used through decoupled code
* app.js: object that controls global state of app
* screen/entry.js: first component that renders an emtpy spinner indicating loading data
* utils/error_handling.js: handle error handling logic
* utils/push_notifications: handle push notification logic

**Other Changes that impact cold startup time**
1. disable multidex
2. stripped localization files, will require a different solution (possibly a native solution)
3. cache native module constants on cold startup (files included in this PR but not utilized)
   * Check `native_changes/` directory
   * Uses SharedPreference to store native module constants

#### Device Information
This PR was tested on: [Android Nexus 5 Emulator, iPhone 6, Android Nexus 6P release mode] 

#### Change in Action
ezgif com-gif-maker 3

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -71,10 +71,11 @@ post-install:
 	@rm -f node_modules/intl/.babelrc
 	@# Hack to get react-intl and its dependencies to work with react-native
 	@# Based off of https://github.com/este/este/blob/master/gulp/native-fix.js
-	@sed -i'' -e 's|"./locale-data/index.js": false|"./locale-data/index.js": "./locale-data/index.js"|g' node_modules/react-intl/package.json
-	@sed -i'' -e 's|"./lib/locales": false|"./lib/locales": "./lib/locales"|g' node_modules/intl-messageformat/package.json
-	@sed -i'' -e 's|"./lib/locales": false|"./lib/locales": "./lib/locales"|g' node_modules/intl-relativeformat/package.json
-	@sed -i'' -e 's|"./locale-data/complete.js": false|"./locale-data/complete.js": "./locale-data/complete.js"|g' node_modules/intl/package.json
+	#@sed -i'' -e 's|"./locale-data/index.js": false|"./locale-data/index.js": "./locale-data/index.js"|g' node_modules/react-intl/package.json
```
Comment (miguelespinoza) [inline Makefile]: Will be removed for MM PR

File: app/app.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+        const shouldStart = await handleManagedConfig();
+        if (shouldStart) {
+            this.setManagedConfig(shouldStart);
+            this.startApp();
+        }
+    };
+
+    startApp = () => {
+        // TODO: should we check if entryComponent exists?
```
Comment (miguelespinoza) [inline app/app.js]: Still a concern, I've tested on cold startup through the launcher and push notification and this would not be necessary. 
 
I added this comment to check if there's another way a user can end up attempting to launchApp and not have reference to the Entry component.  Resulting in the app never starting

File: app/components/start/empty_toolbar.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+const {
+    ANDROID_TOP_LANDSCAPE,
+    ANDROID_TOP_PORTRAIT,
+    IOS_TOP_LANDSCAPE,
+    IOS_TOP_PORTRAIT,
+    STATUS_BAR_HEIGHT
+} = ViewTypes;
+
+export default class ChannelNavBar extends PureComponent {
```
Comment (miguelespinoza) [inline app/components/start/empty_toolbar.js]: 1. Fix name of class to EmptyToolbar
2. Add PropTypes support

File: app/i18n/index.js [hunk 4]
```diff
@@ -2,69 +2,14 @@
...
-addLocaleData(itLocaleData);
-addLocaleData(jaLocaleData);
-addLocaleData(koLocaleData);
-addLocaleData(nlLocaleData);
-addLocaleData(plLocaleData);
-addLocaleData(ptLocaleData);
-addLocaleData(trLocaleData);
-addLocaleData(ruLocaleData);
-addLocaleData(zhLocaleData);
-
```
Comment (miguelespinoza) [inline app/i18n/index.js]: These changes will be reverted back to original, in favor of a native solution.

More than likely will be another PR

File: app/utils/sentry/middleware.js [hunk 5]
```diff
@@ -10,7 +10,7 @@ export function createSentryMiddleware() {
     return (store) => { // eslint-disable-line no-unused-vars
         return (next) => {
             return (action) => {
-                Sentry.captureBreadcrumb(makeBreadcrumbFromAction(action));
+                // Sentry.captureBreadcrumb(makeBreadcrumbFromAction(action));
```
Comment (miguelespinoza) [inline app/utils/sentry/middleware.js]: Forgot to uncomment these

File: index.js [hunk 6]
```diff
@@ -4,11 +4,13 @@
...
 import ShareExtension from 'share_extension/android';
 
 if (Platform.OS === 'android') {
     AppRegistry.registerComponent('MattermostShare', () => ShareExtension);
 }
 
-const app = new Mattermost();
+if (Platform.OS === 'ios') {
+    const app = new Mattermost();
+}
```
Comment (miguelespinoza) [inline index.js]: I'm conflicted by this.   The PR addresses cold startup time for Android yet it makes core changes that would affect both Android and iOS.

The changes made are iOS compliant for the most part some UI cosmetic changes.  My concern is around push notification logic.  The logic has been ported yet it would need throughout testing to ensure there's no regression.

Since it would require creating a release build I decided just to keep original functionality for iOS.

We'll look at a proper solution for iOS in the next couple of days

File: native_changes/RNDeviceModule.java [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+
+        constants.put("phoneNumber", preferences.getString("phoneNumber", ""));
+        constants.put("carrier", preferences.getString("carrier", ""));
+        constants.put("maxMemory", preferences.getLong("maxMemory", 0));
+        constants.put("totalMemory", preferences.getLong("totalMemory", 0));
+
+        return constants;
+    }
+
+    private void saveConstants(Map<String, Object> constants, Context ctx) {
```
Comment (miguelespinoza) [inline native_changes/RNDeviceModule.java]: Saves constants into Android's key/value store. 

DeviceInfo, in particular, creates an overhead to initialize:

```
constants.put("userAgent", WebSettings.getDefaultUserAgent(this.reactContext));
```

This is a constant we know won't change in a device therefore we can cache without impacting functionality in the app

File: app/app.js [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        if (shouldStartCache && !shouldStart) {
+            this.setManagedConfig(false);
+            mattermostManaged.quitApp();
+            return;
+        }
+
+        this.setManagedConfig(true);
+    };
+
+    launchApp = async () => {
```
Comment (miguelespinoza) [inline app/app.js]: launchApp logic so it can be used throughout decoupled code

File: app/screens/entry/entry.js [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+            this.setReplyNotifications();
+
+            if (Platform.OS === 'android') {
+                this.launchForAndroid();
+                return;
+            }
+
+            this.launchForiOS();
+        }
+    };
```
Comment (miguelespinoza) [inline app/screens/entry/entry.js]: At this point the only useful thing Entry component does is wait for `rehydration`

That's something that may move to `mattermost.android.js` in order to subscribe to the store sooner.  Before the component even renders, there's an overhead to start a new Activity so that may be a good next change.

For the meantime, it works just fine in this class

File: app/mattermost.android.js [hunk 10]
```diff
@@ -0,0 +1,436 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import 'babel-polyfill';
```
Comment (miguelespinoza) [inline app/mattermost.android.js]: @jarredwitt This may be hard to follow but this is the basic anatomy of `mattemost.android.js`

1. load heavy modules in a lazy manner
2. initializeModules function ready to be called after Channel has been launched.
   * These modules do not depend on cold startup, so they can be initialized after the Channel has rendered
3. contain global listeners/callbacks in this file so it can maintain reference throught the app's existence `mattermost.js` class
4. start `Entry` screen, notice not much has been initialized

After this we do the same thing wait for `redux rehydration`

File: app/screens/index.js [hunk 11]
```diff
@@ -67,48 +21,49 @@ function wrapWithContextProvider(Comp, excludeEvents = true) {
...
+    Navigation.registerComponent('Settings', () => wrapWithContextProvider(require('app/screens/settings/general').default), store, Provider); //eslint-disable-line global-require
+    Navigation.registerComponent('SSO', () => wrapWithContextProvider(require('app/screens/sso').default), store, Provider); //eslint-disable-line global-require
+    Navigation.registerComponent('Table', () => wrapWithContextProvider(require('app/screens/table').default), store, Provider); //eslint-disable-line global-require
+    Navigation.registerComponent('TableImage', () => wrapWithContextProvider(require('app/screens/table_image').default), store, Provider); //eslint-disable-line global-require
+    Navigation.registerComponent('Thread', () => wrapWithContextProvider(require('app/screens/thread').default), store, Provider); //eslint-disable-line global-require
+    Navigation.registerComponent('UserProfile', () => wrapWithContextProvider(require('app/screens/user_profile').default), store, Provider); //eslint-disable-line global-require
```
Comment (miguelespinoza) [inline app/screens/index.js]: This will be merged by another PR in MM, so I'll remove for PR back to MM master

Comment (miguelespinoza) [inline app/screens/index.js]: After rebase with latest changes it bumped bundle to 4MB which also bumped native cold startup by a couple 100 milliseconds.

It's inevitable that adding new JS code/screens will introduce cold startup regressions.
One thing we can do is investigate `react-native unbundle`.  

Also, a thought I had was to break registerScreens into registerIntialScreens and registerAdditionalScreens

**registerInitialScreens** would register components that are required by the cold startup process.
- Root
- Entry
- SelectServer
- Login 
- SSO
- Channel

**registerAdditionalScreens** would register everything else.  Maybe can register in Channel component 🤷‍♂️ 
Would be an easy change to make to see if this helps...

File: app/mattermost.android.js [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+
+import App from './app';
+
+const AUTHENTICATION_TIMEOUT = 5 * 60 * 1000;
+
+export const app = new App();
+export const store = configureStore(initialState);
+registerScreens(store, Provider);
+
+if (Platform.OS === 'android') {
```
Comment (jarredwitt) [inline app/mattermost.android.js]: Is this needed since the file has an .android.js extension?

Comment (miguelespinoza) [reply app/mattermost.android.js]: So the ultimate goal is to support one file `mattermost.js` instead of 2 based on os.

I kept the old mattermost.js file (converted to .ios.js) to support iOS in the meantime while we come up with a plan for testing iOS
